.github/workflows/validate-pr.yml

```
@@ -49,14 +49,20 @@ jobs:
     runs-on: ubuntu-20.04
     env:
       PR_TITLE: ${{ github.event.pull_request.title }}
+      TARGET_BRANCH: ${{ github.event.pull_request.base.ref }}
     steps:
       # Backports or PR that target a release branch directly should mention the target branch in the title, for example:
       # [X.Y backport] Some change that needs backporting to X.Y
       # [X.Y] Change directly targeting the X.Y branch
       - name: Get branch from PR title
         id: title_branch
-        run: echo "$PR_TITLE" | sed -n 's/^\[\([0-9]*\.[0-9]*\)[^]]*\].*/branch=\1/p' >> $GITHUB_OUTPUT
+        run: |
+          # get the intended major version prefix ("[27.1 backport]" -> "27.") from the PR title.
+          echo "$PR_TITLE" | sed -n 's/^\[\([0-9]*\.\)[^]]*\].*/branch=\1/p' >> $GITHUB_OUTPUT
+
+          # get major version prefix from the release branch ("27.x -> "27.")
+          echo "$TARGET_BRANCH" | sed -n 's/^\([0-9]*\.\).*/target_branch=\1/p' >> $GITHUB_OUTPUT
```

I wonder if using [the regex match construct in bash](https://www.gnu.org/software/bash/manual/html_node/Conditional-Constructs.html#:~:text=An%20additional%20binary,it%20does%20not.) would make this more readable, or at least slightly less magical? At the very least, it would allow for the use of simpler regexes that don't have to match the entire input in order to suppress unintended output.
```suggestion
          # get the intended major version prefix ("[27.1 backport]" -> "27.") from the PR title.
          [[ "$PR_TITLE" =~ ^\[\([0-9]*\.\)[^]]*\] ]] && echo "branch=$BASH_REMATCH[1]" >> $GITHUB_OUTPUT

          # get major version prefix from the release branch ("27.x -> "27.")
          [[ "$TARGET_BRANCH" =~ ^\([0-9]*\.\) ]] && echo "target_branch=$BASH_REMATCH[1]" >> $GITHUB_OUTPUT
```

---

Do actions use bash as default, or is it POSIX shell by default? I was even contemplating if we perhaps should combine some of those tests and just create a basic go binary / script for some of them. I agree it's somewhat ugly currently (but hey, it works!), but even with your suggestions it's still somewhat ugly, and wondering if we can avoid all those indirects of;

- put branch name into env-var
- parse env-var, and redirect to `$GITHUB_OUTPUT`
- in order to have it available again in GHA
- only to check if it matches
- then printing the error

It's a lot of indirects, which we possibly could all simplify with some other means 🤔

---

https://github.com/actions/github-script?

---

> Do actions use bash as default, or is it POSIX shell by default?

Bash as default, if available. https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idstepsshell

---

> https://github.com/actions/github-script?

LOL, yeah, I was trying to avoid adding TypeScript to my list of things to get good at 😂

---

> It's a lot of indirects, which we possibly could all simplify with some other means 🤔

_Very_ good point! This could all be done from a single workflow step. And we don't even need to stuff the branch name into an env var!

```suggestion
          # get the intended major version prefix ("[27.1 backport]" -> "27.") from the PR title.
          [[ "$PR_TITLE" =~ ^\[\([0-9]*\.\)[^]]*\] ]] && branch="${BASH_REMATCH[1]}"
          
          # get major version prefix from the release branch ("27.x -> "27.")
          [[ "$GITHUB_BASE_REF" =~ ^\([0-9]*\.\) ]] && target_branch="${BASH_REMATCH[1]}"
          
          if [[ "$GITHUB_BASE_REF" != "$branch" ]] && ! [[ "$GITHUB_BASE_REF" == "master" && "$branch" == "" ]]; then
              echo "::error::PR is opened against the $GITHUB_BASE_REF branch, but its title suggests otherwise."
              exit 1
          fi
```